### PR TITLE
Implement turn timer feature

### PR DESCRIPTION
**Pull request by Nordsoft91**
Design suggestions are welcomed. Would appreciate for testing as many bugs could be potentially found.
What is covered:
 - timer if frozen during battles
 - timer is not frozen on queries, but turn pass won't happen until all queries resolved

Known problem: timer is not frozen on turn pass in hotseat mode. This needs to be done separately because change and impact of supporting it is quite big.

### CODE REVIEW

**Comment by IvanSavenko** [comment]
This is what we actually need right now.
H3-like timers are not used neither in SP nor in MP
https://github.com/vcmi/vcmi/issues/1364

**Comment by nullkiller** [comment]
Does AI adequately react on it?

File: client/adventureMap/AdventureMapInterface.cpp [hunk 1]
```diff
@@ -61,6 +63,9 @@ AdventureMapInterface::AdventureMapInterface():
 	shortcuts->setState(EAdventureState::MAKING_TURN);
 	widget->getMapView()->onViewMapActivated();
 
+	if(LOCPLINT->cb->getStartInfo()->turnTime > 0)
+		watches = std::make_shared<TurnTimerWidget>();
+
```
Comment (IvanSavenko) [inline client/adventureMap/AdventureMapInterface.cpp:68]: Should be defined in widget / configurable UI

File: server/CGameHandler.cpp [hunk 2]
```diff
@@ -2075,10 +2083,31 @@ void CGameHandler::run(bool resume)
 			if(playerColor != PlayerColor::CANNOT_DETERMINE)
 			{
 				//wait till turn is done
+				const int waitTime = 100; //ms
+				int turnTimePropagateFrequency = 5000; //do not send updates too frequently
 				boost::unique_lock<boost::mutex> lock(states.mx);
 				while(states.players.at(playerColor).makingTurn && lobby->state == EServerState::GAMEPLAY)
 				{
-					static time_duration p = milliseconds(100);
+					if(gs->getStartInfo()->turnTime > 0 && !gs->curB) //turn timer check
```
Comment (IvanSavenko) [inline server/CGameHandler.cpp:2091]: Please don't add any new functionality to CGameHandler, that class is already unmanageable and I want to split it to pieces soon.

Move this to a new class, let's call it TurnTimerProcessor for example.

File: client/adventureMap/TurnTimerWidget.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+#include "../render/Colors.h"
+#include "../render/EFont.h"
+#include "../gui/CGuiHandler.h"
+#include "../gui/TextAlignment.h"
+#include "../widgets/Images.h"
+#include "../widgets/TextControls.h"
+#include "../../CCallback.h"
+#include "../../lib/CPlayerState.h"
+
+#include <SDL_render.h>
```
Comment (IvanSavenko) [inline client/adventureMap/TurnTimerWidget.cpp:24]: Please don't use SDL directly in UI code

**Review by IvanSavenko — COMMENTED** [review]
- Needs chess timers, not simple countdown
- Needs AI handling 
- Needs human handling (potentially including closing all active dialogs)

**Comment by dydzio0614** [comment]
@Nordsoft91 since this is considered one of most important features for 1.4 release, ensure that you have time to polish it until 1.4 releases, otherwise I would probably prefer Ivan implementing it, including simultaneous turns support

**Comment by Nordsoft91** [comment]
> @Nordsoft91 since this is considered one of most important features for 1.4 release, ensure that you have time to polish it until 1.4 releases, otherwise I would probably prefer Ivan implementing it, including simultaneous turns support for timers. There are bunch of less convoluted features that can still be done, like AI cheating (#1931) or spells usage on/off toggle for autobattle (got a lot of upvotes in discord when mentioned) - at least they are less prone to major changes when simultaneous turns come

quite weird preference, because collaboration and labor division usually speeds up the development, also I don't see interference with simultaneous turns. As for time allocation, if there will be regression surely will do the best to fix it.
Unless there are better/cheaper/faster incoming solutions I'll keep PR

**Comment by Nordsoft91** [comment]
> Does AI adequately react on it?

I tested few times turn expiration for AI and it seems to be fine

> Needs human handling (potentially including closing all active dialogs)

works from the box, except scenario I described in description

**Comment by IvanSavenko** [comment]
> works from the box, except scenario I described in description

- No notification of any kind. When you are time-pressed you won't be looking on countdown. 
- If your hero is moving at the end of your turn you start getting server errors.
- If your hero is flying over obstacles he will stop movement while on top of them.

**Comment by IvanSavenko** [comment]
And another reminder that we need #1364 and not this version

**Comment by krs0** [comment]
On audio notifications. Besides the 10 seconds countdown, would a single beep at 30 sec be helpful?

**Comment by Nordsoft91** [comment]
@krs0 , I think possible to make different sound configuration, do you have sound we may use?

**Comment by dydzio0614** [comment]
maybe https://www.youtube.com/watch?v=aJwZntohTt8 :rofl:

**Comment by krs0** [comment]
@dydzio0614 Haha :)

I am not good at this kind of stuff, but maybe better than hat dydzio provided :)

Default system message (sysmsg I think) maybe would scare me that I am in last 10 seconds.

So maybe Something different like:
https://notificationsounds.com/message-tones/that-was-quick-606
Or
https://notificationsounds.com/message-tones/good-news-387

File: client/gui/InterfaceObjectConfigurable.cpp [hunk 4]
```diff
@@ -190,18 +190,29 @@ ColorRGBA InterfaceObjectConfigurable::readColor(const JsonNode & config) const
...
-			return Colors::GREEN;
-		if(config.String() == "orange")
-			return Colors::ORANGE;
-		if(config.String() == "bright-yellow")
-			return Colors::BRIGHT_YELLOW;
+		if(config.isString())
+		{
+			if(config.String() == "yellow")
+				return Colors::YELLOW;
+			if(config.String() == "white")
```
Comment (krs0) [inline client/gui/InterfaceObjectConfigurable.cpp:197]: One of the things I hate about C++: a simple thing like using string values in enum, has no support in language :(

Could this be done separately? Where the enum is defined?

A map of some sorts 
Or a method to enum class...whith the switches inside
FromString(const std:string &enumValue)

Comment (IvanSavenko) [reply client/gui/InterfaceObjectConfigurable.cpp:197]: This is not enum, but a set of static constants. I guess in theory we also add fromString method in there, but so far this is the only place where we use string->color conversion so probably no need.

**Comment by krs0** [comment]
BUG: Restart scenario does not restart timer.

**Comment by krs0** [comment]
At start of turn would'nt it be better to be actual set timer value? 00 jumping to actual timer value is a little confusing.

image

**Comment by krs0** [comment]
In hot seat mode timer runs while turn is not started yet. Should not count down on this screen.
image

**Comment by krs0** [comment]
I can stay forever in a mine :)

image

**Comment by IvanSavenko** [comment]
Tested AI - game client hangs when AI fails to end turn in time. Reproduced twice out of two attempts

**Comment by IvanSavenko** [comment]
Also, perhaps use player color as color for timer text? So red text when red is making turn, blue when blue is making turn, etc.
This is how hota/hd does it. Right now it is a bit weird to see timer during AI turn.

**Comment by Nordsoft91** [comment]
> Tested AI - game client hangs when AI fails to end turn in time. Reproduced twice out of two attempts

Failed to fix, so disabled time expiration for AI players

> Also, perhaps use player color as color for timer text? 

Done

**Comment by Nordsoft91** [comment]
@krs0 , visual bugs should be fixed, but auto-response for queries (hero exchange, garrisons, etc.) is way more complex.
I implemented it in a way, that time continue expiring while you on query dialog but turn cannot be end. Changing this requires significant refactoring of multiple components so I decide to postpone it.

**Comment by Nordsoft91** [comment]
Merge now to avoid further complex conflict resolution. Need more testing, UI and query resolution

File: server/CGameHandler.cpp [hunk 5]
```diff
@@ -1048,10 +1060,15 @@ void CGameHandler::run(bool resume)
...
+				const int waitTime = 100; //ms
 				boost::unique_lock<boost::mutex> lock(states.mx);
 				while(states.players.at(playerColor).makingTurn && lobby->state == EServerState::GAMEPLAY)
 				{
-					static time_duration p = milliseconds(100);
+					turnTimerHandler.onPlayerMakingTurn(gs->players[playerColor], waitTime);
+					if(gs->curB)
+						turnTimerHandler.onBattleLoop(waitTime);
+
+					static time_duration p = milliseconds(waitTime);
```
[the remark this answers is not in the document]
Comment (IvanSavenko) [reply server/CGameHandler.cpp:1071]: Oh. And there is no concurrency protection!
Yes, there is `states.mx` mutex here, but this mutex only protects `states` member.

Game currently processes packs in `CVCMIServer::threadHandleClient`.
But this code runs in main thread

File: server/CGameHandler.cpp [hunk 6]
```diff
@@ -1248,7 +1268,17 @@ bool CGameHandler::moveHero(ObjectInstanceID hid, int3 dst, ui8 teleporting, boo
...
+		{
+			for(auto topQuery = queries->topQuery(h->tempOwner); true; topQuery = queries->topQuery(h->tempOwner))
+			{
+				moveQuery = std::dynamic_pointer_cast<CHeroMovementQuery>(topQuery);
+				if(moveQuery)
+					queries->popIfTop(moveQuery);
+				else
+					break;
+			}
+		}
```
Comment (IvanSavenko) [inline server/CGameHandler.cpp:1281]: Likely cause for regression  #2658

**Pull request merged**